File: facebook-core/src/main/java/com/facebook/appevents/ml/Operator.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents.ml;

import com.facebook.internal.instrument.crashshield.AutoHandleExceptions;

@AutoHandleExceptions
final class Operator {

  static void addmv(MTensor x, MTensor b) {
    int n_example = x.getShape(0);
    int seq_len = x.getShape(1);
    int input_size = x.getShape(2);
    float[] x_data = x.getData();
    float[] b_data = b.getData();

    for (int i = 0; i < n_example; i++) {
      for (int j = 0; j < seq_len; j++) {
        for (int k = 0; k < input_size; k++) {
          x_data[i * seq_len * input_size + j * input_size + k] += b_data[k];
        }
      }
    }
  }

  static MTensor mul(MTensor x, MTensor w) {
    int n_examples = x.getShape(0);
    int input_size = w.getShape(0);
    int output_size = w.getShape(1);
    MTensor y = new MTensor(new int[] {n_examples, output_size});
    float[] x_data = x.getData();
    float[] w_data = w.getData();
    float[] y_data = y.getData();

    for (int i = 0; i < n_examples; i++) {
      for (int j = 0; j < output_size; j++) {
        y_data[i * output_size + j] = 0;
        for (int k = 0; k < input_size; k++) {
          y_data[i * output_size + j] += x_data[i * input_size + k] * w_data[k * output_size + j];
        }
      }
    }
    return y;
  }

  static void relu(MTensor x) {
    float[] x_data = x.getData();
    for (int i = 0; i < x_data.length; i++) {
      if (x_data[i] < 0) {
        x_data[i] = 0;
      }
    }
  }

  static void flatten(MTensor x, int start_dim) {
    if (start_dim >= x.getShapeSize()) {
      return;
    }
    int output_size = 1;
    for (int i = start_dim; i < x.getShapeSize(); i++) {
      output_size *= x.getShape(i);
    }
    int[] new_shape = new int[start_dim + 1];
    for (int i = 0; i < start_dim; i++) {
      new_shape[i] = x.getShape(i);
    }
    new_shape[start_dim] = output_size;
    x.reshape(new_shape);
  }

  static MTensor concatenate(MTensor[] tensors) {
    int n_examples = tensors[0].getShape(0);
    int output_size = 0;
    for (int i = 0; i < tensors.length; i++) {
      output_size += tensors[i].getShape(1);
    }
    MTensor y = new MTensor(new int[] {n_examples, output_size});
    float[] y_data = y.getData();

    for (int n = 0; n < n_examples; n++) {
      int desPos = n * output_size;
      for (int i = 0; i < tensors.length; i++) {
        float[] x_data = tensors[i].getData();
        int input_size = tensors[i].getShape(1);
        System.arraycopy(x_data, n * input_size, y_data, desPos, input_size);
        desPos += input_size;
      }
    }
    return y;
  }

  static void softmax(MTensor x) {
    int n_examples = x.getShape(0);
    int input_size = x.getShape(1);
    float[] x_data = x.getData();
    for (int n = 0; n < n_examples; n++) {
      int start_idx = n * input_size;
      int end_idx = start_idx + input_size;
      float max = Float.MIN_VALUE;
      float sum = 0;

      for (int i = start_idx; i < end_idx; i++) {
        if (x_data[i] > max) {
          max = x_data[i];
        }
      }

      for (int i = start_idx; i < end_idx; i++) {
        x_data[i] = (float) Math.exp(x_data[i] - max);
      }

      for (int i = start_idx; i < end_idx; i++) {
        sum += x_data[i];
      }

      for (int i = start_idx; i < end_idx; i++) {
        x_data[i] = x_data[i] / sum;
      }
    }
  }

  static MTensor dense(MTensor x, MTensor w, MTensor b) {
    int n_examples = x.getShape(0);
    int output_size = b.getShape(0);
    MTensor y = mul(x, w);
    float[] b_data = b.getData();
    float[] y_data = y.getData();

    for (int i = 0; i < n_examples; i++) {
      for (int j = 0; j < output_size; j++) {
        y_data[i * output_size + j] += b_data[j];
      }
    }
    return y;
  }

  static MTensor embedding(String[] texts, int seq_len, MTensor w) {
    int n_examples = texts.length;
    int embedding_size = w.getShape(1);
    MTensor y = new MTensor(new int[] {n_examples, seq_len, embedding_size});
    float[] y_data = y.getData();
    float[] w_data = w.getData();

    for (int i = 0; i < n_examples; i++) {
      int[] vectorize_text = Utils.vectorize(texts[i], seq_len);
      for (int j = 0; j < seq_len; j++) {
        System.arraycopy(
            w_data,
            vectorize_text[j] * embedding_size,
            y_data,
            embedding_size * seq_len * i + embedding_size * j,
            embedding_size);
      }
    }
    return y;
  }

  static MTensor transpose2D(MTensor x) {
    int m = x.getShape(0);
    int n = x.getShape(1);
    MTensor y = new MTensor((new int[] {n, m}));
    float[] x_data = x.getData();
    float[] y_data = y.getData();

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        y_data[j * m + i] = x_data[i * n + j];
      }
    }
    return y;
  }

  static MTensor transpose3D(MTensor x) {
    int m = x.getShape(0);
    int n = x.getShape(1);
    int p = x.getShape(2);
    MTensor y = new MTensor(new int[] {p, n, m});
    float[] x_data = x.getData();
    float[] y_data = y.getData();

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        for (int k = 0; k < p; k++) {
          y_data[k * m * n + j * m + i] = x_data[i * n * p + j * p + k];
        }
      }
    }
    return y;
  }

  static MTensor conv1D(MTensor x, MTensor w) {
    int n_examples = x.getShape(0);
    int input_seq_len = x.getShape(1);
    int input_size = x.getShape(2);
    int kernel_size = w.getShape(0);
    int output_seq_len = input_seq_len - kernel_size + 1;
    int output_size = w.getShape(2);
    MTensor y = new MTensor(new int[] {n_examples, output_seq_len, output_size});
    float[] x_data = x.getData();
    float[] y_data = y.getData();
    float[] w_data = w.getData();

    for (int n = 0; n < n_examples; n++) {
      for (int o = 0; o < output_size; o++) {
        for (int i = 0; i < output_seq_len; i++) {
          float sum = 0;
          for (int m = 0; m < kernel_size; m++) {
            for (int k = 0; k < input_size; k++) {
              sum +=
                  x_data[n * (input_seq_len * input_size) + (m + i) * input_size + k]
                      * w_data[(m * input_size + k) * output_size + o];
            }
          }
          y_data[(n * (output_seq_len * output_size) + i * output_size + o)] = sum;
        }
      }
    }
    return y;
  }

  static MTensor maxPool1D(MTensor x, int pool_size) {
    int n_examples = x.getShape(0);
    int input_seq_len = x.getShape(1);
    int input_size = x.getShape(2);
    int output_seq_len = input_seq_len - pool_size + 1;
    MTensor y = new MTensor(new int[] {n_examples, output_seq_len, input_size});
    float[] x_data = x.getData();
    float[] y_data = y.getData();

    for (int n = 0; n < n_examples; n++) {
      for (int c = 0; c < input_size; c++) {
        for (int i = 0; i < output_seq_len; i++) {
          int y_index = n * output_seq_len * input_size + i * input_size + c;
          int x_index = n * input_seq_len * input_size + i * input_size + c;
          y_data[y_index] = Float.MIN_VALUE;
          for (int r = 0; r < pool_size; r++) {
            y_data[y_index] = Math.max(y_data[y_index], x_data[x_index + r * input_size]);
          }
        }
      }
    }
    return y;
  }
}

File: facebook-core/src/main/java/com/facebook/appevents/ml/Operator.kt
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.facebook.appevents.ml

import com.facebook.internal.instrument.crashshield.AutoHandleExceptions
import kotlin.math.exp
import kotlin.math.max

@AutoHandleExceptions
internal object Operator {
  @JvmStatic
  fun addmv(x: MTensor, b: MTensor) {
    val exampleSize = x.getShape(0)
    val seqLength = x.getShape(1)
    val inputSize = x.getShape(2)
    val xData = x.data
    val bData = b.data
    for (i in 0 until exampleSize) {
      for (j in 0 until seqLength) {
        for (k in 0 until inputSize) {
          xData[i * seqLength * inputSize + j * inputSize + k] += bData[k]
        }
      }
    }
  }

  @JvmStatic
  fun mul(x: MTensor, w: MTensor): MTensor {
    val exampleSize = x.getShape(0)
    val inputSize = w.getShape(0)
    val outputSize = w.getShape(1)
    val y = MTensor(intArrayOf(exampleSize, outputSize))
    val xData = x.data
    val wData = w.data
    val yData = y.data
    for (i in 0 until exampleSize) {
      for (j in 0 until outputSize) {
        yData[i * outputSize + j] = 0f
        for (k in 0 until inputSize) {
          yData[i * outputSize + j] += xData[i * inputSize + k] * wData[k * outputSize + j]
        }
      }
    }
    return y
  }

  @JvmStatic
  fun relu(x: MTensor) {
    val xData = x.data
    for (i in xData.indices) {
      if (xData[i] < 0) {
        xData[i] = 0f
      }
    }
  }

  @JvmStatic
  fun flatten(x: MTensor, startDim: Int) {
    if (startDim >= x.shapeSize) {
      return
    }
    var outputSize = 1
    for (i in startDim until x.shapeSize) {
      outputSize *= x.getShape(i)
    }
    val newShape = IntArray(startDim + 1)
    for (i in 0 until startDim) {
      newShape[i] = x.getShape(i)
    }
    newShape[startDim] = outputSize
    x.reshape(newShape)
  }

  @JvmStatic
  fun concatenate(tensors: Array<MTensor>): MTensor {
    val exampleSize = tensors[0].getShape(0)
    var outputSize = 0
    for (i in tensors.indices) {
      outputSize += tensors[i].getShape(1)
    }
    val y = MTensor(intArrayOf(exampleSize, outputSize))
    val yData = y.data
    for (n in 0 until exampleSize) {
      var desPos = n * outputSize
      for (i in tensors.indices) {
        val xData = tensors[i].data
        val inputSize = tensors[i].getShape(1)
        System.arraycopy(xData, n * inputSize, yData, desPos, inputSize)
        desPos += inputSize
      }
    }
    return y
  }

  @JvmStatic
  fun softmax(x: MTensor) {
    val exampleSize = x.getShape(0)
    val inputSize = x.getShape(1)
    val xData = x.data
    for (n in 0 until exampleSize) {
      val startIndex = n * inputSize
      val endIndex = startIndex + inputSize
      var max = Float.MIN_VALUE
      var sum = 0f
      for (i in startIndex until endIndex) {
        if (xData[i] > max) {
          max = xData[i]
        }
      }
      for (i in startIndex until endIndex) {
        xData[i] = exp((xData[i] - max).toDouble()).toFloat()
        sum += xData[i]
      }
      for (i in startIndex until endIndex) {
        xData[i] = xData[i] / sum
      }
    }
  }

  @JvmStatic
  fun dense(x: MTensor, w: MTensor, b: MTensor): MTensor {
    val exampleSize = x.getShape(0)
    val outputSize = b.getShape(0)
    val y = mul(x, w)
    val bData = b.data
    val yData = y.data
    for (i in 0 until exampleSize) {
      for (j in 0 until outputSize) {
        yData[i * outputSize + j] += bData[j]
      }
    }
    return y
  }

  @JvmStatic
  fun embedding(texts: Array<String>, seqLength: Int, w: MTensor): MTensor {
    val exampleSize = texts.size
    val embeddingSize = w.getShape(1)
    val y = MTensor(intArrayOf(exampleSize, seqLength, embeddingSize))
    val yData = y.data
    val wData = w.data
    for (i in 0 until exampleSize) {
      val vectorizedText = Utils.vectorize(texts[i], seqLength)
      for (j in 0 until seqLength) {
        System.arraycopy(
            wData,
            vectorizedText[j] * embeddingSize,
            yData,
            embeddingSize * seqLength * i + embeddingSize * j,
            embeddingSize)
      }
    }
    return y
  }

  @JvmStatic
  fun transpose2D(x: MTensor): MTensor {
    val m = x.getShape(0)
    val n = x.getShape(1)
    val y = MTensor(intArrayOf(n, m))
    val xData = x.data
    val yData = y.data
    for (i in 0 until m) {
      for (j in 0 until n) {
        yData[j * m + i] = xData[i * n + j]
      }
    }
    return y
  }

  @JvmStatic
  fun transpose3D(x: MTensor): MTensor {
    val m = x.getShape(0)
    val n = x.getShape(1)
    val p = x.getShape(2)
    val y = MTensor(intArrayOf(p, n, m))
    val xData = x.data
    val yData = y.data
    for (i in 0 until m) {
      for (j in 0 until n) {
        for (k in 0 until p) {
          yData[k * m * n + j * m + i] = xData[i * n * p + j * p + k]
        }
      }
    }
    return y
  }

  @JvmStatic
  fun conv1D(x: MTensor, w: MTensor): MTensor {
    val exampleSize = x.getShape(0)
    val inputSeqLength = x.getShape(1)
    val inputSize = x.getShape(2)
    val kernelSize = w.getShape(0)
    val outputSeqLength = inputSeqLength - kernelSize + 1
    val outputSize = w.getShape(2)
    val y = MTensor(intArrayOf(exampleSize, outputSeqLength, outputSize))
    val xData = x.data
    val yData = y.data
    val wData = w.data
    for (n in 0 until exampleSize) {
      for (o in 0 until outputSize) {
        for (i in 0 until outputSeqLength) {
          var sum = 0f
          for (m in 0 until kernelSize) {
            for (k in 0 until inputSize) {
              sum +=
                  xData[n * (inputSeqLength * inputSize) + (m + i) * inputSize + k] *
                      wData[(m * inputSize + k) * outputSize + o]
            }
          }
          yData[n * (outputSeqLength * outputSize) + i * outputSize + o] = sum
        }
      }
    }
    return y
  }

  @JvmStatic
  fun maxPool1D(x: MTensor, poolSize: Int): MTensor {
    val exampleSize = x.getShape(0)
    val inputSeqLength = x.getShape(1)
    val inputSize = x.getShape(2)
    val outputSeqLength = inputSeqLength - poolSize + 1
    val y = MTensor(intArrayOf(exampleSize, outputSeqLength, inputSize))
    val xData = x.data
    val yData = y.data
    for (n in 0 until exampleSize) {
      for (c in 0 until inputSize) {
        for (i in 0 until outputSeqLength) {
          val yIndex = n * outputSeqLength * inputSize + i * inputSize + c
          val xIndex = n * inputSeqLength * inputSize + i * inputSize + c
          yData[yIndex] = Float.MIN_VALUE
          for (r in 0 until poolSize) {
            yData[yIndex] = max(yData[yIndex], xData[xIndex + r * inputSize])
          }
        }
      }
    }
    return y
  }
}
